### PR TITLE
NFT Updatable Metadata Extension

### PR DESCRIPTION
**Pull request by clbrge**
This draft EIP adds an optional extension to ERC-721 or ERC-1155 interface to allow controlled metadata updates.

### CODE REVIEW

**Comment by clbrge** [comment]
Let me know if you want me to change the number to 5185...

**Review by aekzacx1997 — APPROVED** [review]

File: EIPS/eip-draft_nft_updatable.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
+---
+eip: ????
```
Comment (MicahZoltu) [inline EIPS/eip-draft_nft_updatable.md:2]:
```suggestion
eip: 5185
```
Don't forget to rename the file to `eip-5185.md` as well.

File: EIPS/eip-draft_nft_updatable.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+Smart contracts should be careful and conscious of using this extension and still allow the metadata URI to be updated in some contexts (by not having the same URI returned by `tokenURI` or `uri` for a given tokenId). They need to take into account if previous changes could have been already broadcasted for that NFT, if these changes are compatible with the new "original metadata" and what semantic they decide to associate by combining these two kinds of "updates".
+
+## Backwards Compatibility
+
+The proposal is fully compatible with both ERC-721 and ERC-1155. Third party applications that don't support Updatable Metadata Extension will still be able to use the original metadata for each NFT.
+
+## Gas efficiency
+
+The proposal is extremely gas efficient, since gas costs are only proportional to the frequency of committing changes. Many changes for many tokens can be batched in one transaction for the cost of only one `emit`.
+
```
Comment (MicahZoltu) [inline EIPS/eip-draft_nft_updatable.md:230]: This is not a valid EIP section.  Please move this content as a subsection of a section defined in https://raw.githubusercontent.com/ethereum/EIPs/master/eip-template.md

File: EIPS/eip-draft_nft_updatable.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+      {"tokenId":"1","action":"levelUp"},
+      {"tokenId":"3","action":"levelUp"}
+    ]
+}
+ ```
+
+## Caveats
+
+Smart contracts should be careful and conscious of using this extension and still allow the metadata URI to be updated in some contexts (by not having the same URI returned by `tokenURI` or `uri` for a given tokenId). They need to take into account if previous changes could have been already broadcasted for that NFT, if these changes are compatible with the new "original metadata" and what semantic they decide to associate by combining these two kinds of "updates".
+
```
Comment (MicahZoltu) [inline EIPS/eip-draft_nft_updatable.md:222]: This is not a valid EIP section. Please move this content as a subsection of a section defined in https://raw.githubusercontent.com/ethereum/EIPs/master/eip-template.md

File: EIPS/eip-draft_nft_updatable.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+With `evalString` is found with `recipeKey` in the original metadata recipes list.
+
+If the key is not present in the original metadata list, `previousMetadata` is kept as the valid updated metadata.
+
+If the evaluation throws any errors, `previousMetadata` is kept as the valid updated metadata.
+
+If a validation Schema JSON has been defined and the result JSON `nextMetadata` does not conform, that update is not valid and `previousMetadata` is kept as the valid updated metadata.
+
```
Comment (MicahZoltu) [inline EIPS/eip-draft_nft_updatable.md:159]: This is not a valid EIP section. Please move this content as a subsection of a section defined in https://raw.githubusercontent.com/ethereum/EIPs/master/eip-template.md

**Review by MicahZoltu — COMMENTED** [review]
Security Considerations section is missing and required.

File: EIPS/eip-5185.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+status: Draft
+type: Standards Track
+category: ERC
+requires: 721, 1155
+created: 2022-06-27
+---
+
+## Abstract
+
+This specification defines a standard way to allow controlled NFTs' metadata updates along predefined formulas. Updates of the original metadata are restricted and defined by a set of recipes and the sequence and results of these recipes are deterministic and fully verifiable with on-chain metadata updates event. The proposal depends on and extends the [ERC-721 standard](./eip-721.md) and [ERC-1155 standard](./eip-1155.md).
```
Comment (Pandapip1) [inline EIPS/eip-5185.md:16]:
```suggestion
This specification defines a standard way to allow controlled NFTs' metadata updates along predefined formulas. Updates of the original metadata are restricted and defined by a set of recipes and the sequence and results of these recipes are deterministic and fully verifiable with on-chain metadata updates event. The proposal depends on and extends the [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md).
```

File: EIPS/eip-5185.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+As examples, an NFT representing a house could only allow append-only updates to the list of successive owners, and a game using NFT characters could let some attributes change from time to time (e.g. health, experience, level, etc) while some other would be guaranteed to never change (e.g. physicals traits etc).
+
+This standard extension is compatible with NFTs bridged between Ethereum and L2 networks and allows efficient caching solutions.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+The **metadata updates extension** is OPTIONAL for [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md) contracts.
```
Comment (Pandapip1) [inline EIPS/eip-5185.md:34]:
```suggestion
The **metadata updates extension** is OPTIONAL for [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md) contracts.
```

File: EIPS/eip-5185.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+If the key is not present in the original metadata list, `previousMetadata` is kept as the valid updated metadata.
+
+If the evaluation throws any errors, `previousMetadata` is kept as the valid updated metadata.
+
+If a validation Schema JSON has been defined and the result JSON `nextMetadata` does not conform, that update is not valid and `previousMetadata` is kept as the valid updated metadata.
+
+## Rationale
+
+There have been numerous interesting uses of ERC-721 and ERC-1155 smart contracts that associate for each token essential and significant metadata. While some projects (e.g. EtherOrcs) have experimented successfully to manage these metadata on-chain, that experimental solution will always be limited by the cost and speed of generating and storing JSON on-chain. Symmetrically, while storing the JSON metadata at URI endpoint controlled by traditional servers permit limitless updates the the metadata for each NFT, it is somehow defeating in many uses cases, the whole purpose of using a trustless blockchain to manage NFT: indeed users may want or demand more permanence and immutability from the metadata associated with their NFT.
```
Comment (Pandapip1) [inline EIPS/eip-5185.md:162]:
```suggestion
There have been numerous interesting uses of [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md) smart contracts that associate for each token essential and significant metadata. While some projects (e.g. EtherOrcs) have experimented successfully to manage these metadata on-chain, that experimental solution will always be limited by the cost and speed of generating and storing JSON on-chain. Symmetrically, while storing the JSON metadata at URI endpoint controlled by traditional servers permit limitless updates the the metadata for each NFT, it is somehow defeating in many uses cases, the whole purpose of using a trustless blockchain to manage NFT: indeed users may want or demand more permanence and immutability from the metadata associated with their NFT.
```
Nit

File: EIPS/eip-5185.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+Most use cases have chosen intermediate solutions like IPFS or arweave to provide some permanence or partial/full immutability of metadata. This is a good solution when an NFT represents a static asset whose characteristics are by nature permanent and immutable (like in the art world) but less so with other use cases like gaming or NFT representing a deed or title. Distinguishable assets in a game often should be allowed to evolve and change over time in a controlled way and titles need to record real life changes.
+
+The advantages of this standard is precisely to allow these types of controlled transformations over time of each NFT metadata by applying sequential transformations starting with the original metadata and using formulas themselves defined in the original metadata.
+
+The original metadata for a given NFT is always defined as the JSON pointed by the result of `tokenURI` for ERC-721 and function `uri` for ERC-1155.
```
Comment (Pandapip1) [inline EIPS/eip-5185.md:168]:
```suggestion
The original metadata for a given NFT is always defined as the JSON pointed by the result of `tokenURI` for [EIP-721](./eip-721.md) and function `uri` for [EIP-1155](./eip-1155.md).
```
Nit

File: EIPS/eip-5185.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+## Security Considerations
+
+A malicious recipe in the original metadata might be constructed as a DDoS vector for third parties marketplaces and tools that calculate NFT updated JSON metadata. They are encouraged to properly encapsulate software in charge of these calculations and put limits for the engine updates processing.
+
+Smart contracts should be careful and conscious of using this extension and still allow the metadata URI to be updated in some contexts (by not having the same URI returned by `tokenURI` or `uri` for a given tokenId over time). They need to take into account if previous changes could have been already broadcasted for that NFT by the contract, if these changes are compatible with the new "original metadata" and what semantic they decide to associate by combining these two kinds of "updates". 
+
+## Backwards Compatibility
+
+The proposal is fully compatible with both ERC-721 and ERC-1155. Third party applications that don't support Updatable Metadata Extension will still be able to use the original metadata for each NFT.
```
Comment (Pandapip1) [inline EIPS/eip-5185.md:229]:
```suggestion
The proposal is fully compatible with both [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md). Third-party applications that don't support this EIP will still be able to use the original metadata for each NFT.
```
Nit(s)

File: EIPS/eip-5185.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+## Copyright
+
```
Comment (Pandapip1) [inline EIPS/eip-5185.md:232]:
```suggestion
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

**Review by lightclient — APPROVED** [review]

**Pull request merged**